### PR TITLE
New Resource : `azurerm_storage_mover_agent`

### PR DESCRIPTION
**Pull request by sinbai**
Swagger Link: https://github.com/Azure/azure-rest-api-specs/blob/2c9032ddf4afee483d424d714a03cd1b74dba8ac/specification/storagemover/resource-manager/Microsoft.StorageMover/stable/2023-03-01/storagemover.json#L674

Test Results:
image

### CODE REVIEW

File: internal/services/storagemover/storage_mover_agent_resource.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+			if model == nil {
+				return fmt.Errorf("retrieving %s: model was nil", *id)
+			}
+
+			state := StorageMoverAgentResourceModel{
+				Name:           id.AgentName,
+				StorageMoverId: storagemovers.NewStorageMoverID(id.SubscriptionId, id.ResourceGroupName, id.StorageMoverName).ID(),
+			}
+
+			properties := &model.Properties
```
Comment (tombuildsstuff) [inline internal/services/storagemover/storage_mover_agent_resource.go:198]: .. why are we de-referencing this value?

File: internal/services/storagemover/storage_mover_agent_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+	return utils.Bool(resp.Model != nil), nil
+}
+
+func (r StorageMoverAgentTestResource) template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
```
Comment (tombuildsstuff) [inline internal/services/storagemover/storage_mover_agent_resource_test.go:104]: the `provider` block should be in each test, not the template - to allow for overriding it

File: internal/services/storagemover/storage_mover_agent_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+func (r StorageMoverAgentTestResource) complete(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+
+%s
+
+resource "azurerm_storage_mover_agent" "test" {
+  name             = "acctest-sa-%d"
+  storage_mover_id = azurerm_storage_mover.test.id
+  arc_resource_id  = "${azurerm_resource_group.test.id}/providers/Microsoft.HybridCompute/machines/examples-hybridComputeName"
```
Comment (tombuildsstuff) [inline internal/services/storagemover/storage_mover_agent_resource_test.go:158]: we can provision a hybrid machine with the provider, so we should be doing that - using a fake VM is likely to cause errors in the service that the Service Team will need to fix, which'll likely return an error in the future - so we should test this properly

File: internal/services/storagemover/storage_mover_agent_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+func (r StorageMoverAgentTestResource) update(data acceptance.TestData) string {
+	template := r.template(data)
+	return fmt.Sprintf(`
+
+%s
+
+resource "azurerm_storage_mover_agent" "test" {
+  name             = "acctest-sa-%d"
+  storage_mover_id = azurerm_storage_mover.test.id
+  arc_resource_id  = "${azurerm_resource_group.test.id}/providers/Microsoft.HybridCompute/machines/examples-hybridComputeName"
```
Comment (tombuildsstuff) [inline internal/services/storagemover/storage_mover_agent_resource_test.go:174]: (same here)

File: website/docs/r/storage_mover_agent.html.markdown [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name which should be used for this Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `arc_resource_id` - (Required) Specifies the fully qualified resource ID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/storage_mover_agent.html.markdown:41]: we don't use the phrase `Resource ID` in the documentation since it's ambiguous between an Azure Resource ID and the Terraform Resource ID

File: website/docs/r/storage_mover_agent.html.markdown [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+* `arc_vm_uuid` - (Required) Specifies the VM UUID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `storage_mover_id` - (Required) Specifies the ID of the Storage Mover. Changing this forces a new resource to be created.
+
+* `description` - (Optional) A description for the Storage Mover Agent.
```
Comment (tombuildsstuff) [inline website/docs/r/storage_mover_agent.html.markdown:47]:
```suggestion
* `description` - (Optional) Specifies a description for this Storage Mover Agent.
```

File: website/docs/r/storage_mover_agent.html.markdown [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+subcategory: "Storage Mover"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_storage_mover_agent"
+description: |-
+  Manages a Storage Mover Agent.
+---
+
+# azurerm_storage_mover_agent
+
+Manages a Storage Mover Agents.
```
Comment (tombuildsstuff) [inline website/docs/r/storage_mover_agent.html.markdown:11]:
```suggestion
Manages a Storage Mover Agent.
```

File: website/docs/r/storage_mover_agent.html.markdown [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name which should be used for this Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `arc_resource_id` - (Required) Specifies the fully qualified resource ID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `arc_vm_uuid` - (Required) Specifies the VM UUID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `storage_mover_id` - (Required) Specifies the ID of the Storage Mover. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/storage_mover_agent.html.markdown:45]: Presumably?

```suggestion
* `storage_mover_id` - (Required) Specifies the ID of the Storage Mover that this Agent should be connected to. Changing this forces a new resource to be created.
```

File: website/docs/r/storage_mover_agent.html.markdown [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name which should be used for this Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `arc_resource_id` - (Required) Specifies the fully qualified resource ID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
+
+* `arc_vm_uuid` - (Required) Specifies the VM UUID of the Hybrid Compute resource for the Storage Mover Agent. Changing this forces a new resource to be created.
```
Comment (tombuildsstuff) [inline website/docs/r/storage_mover_agent.html.markdown:43]: why is this needed, we provide the Resource ID?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @sinbai 

Thanks for this PR - I've taken a look through and left some comments inline, if we can fix those up then we should be able to take another look here.

Thanks!

File: internal/services/storagemover/storage_mover_agent_resource.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+func (r StorageMoverAgentResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"arc_resource_id": {
```
[the remark this answers is not in the document]
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource.go]: yes, the code has been updated.

File: internal/services/storagemover/storage_mover_agent_resource.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+			model := resp.Model
+			if model == nil {
+				return fmt.Errorf("retrieving %s: model was nil", *id)
+			}
+
+			state := StorageMoverAgentResourceModel{
+				Name:           id.AgentName,
+				StorageMoverId: storagemovers.NewStorageMoverID(id.SubscriptionId, id.ResourceGroupName, id.StorageMoverName).ID(),
+			}
```
[the remark this answers is not in the document]
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource.go]: fixed.

File: internal/services/storagemover/storage_mover_agent_resource.go (hunk 1 above)
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource.go]: Code has been updated.

File: internal/services/storagemover/storage_mover_agent_resource_test.go (hunk 2 above)
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource_test.go]: Code has been updated.

File: internal/services/storagemover/storage_mover_agent_resource_test.go (hunk 3 above)
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource_test.go]: Make sense for me. The test cases have been updated.

File: website/docs/r/storage_mover_agent.html.markdown (hunk 5 above)
Comment (sinbai) [reply website/docs/r/storage_mover_agent.html.markdown]: Got it. Thanks for letting me know this.

File: website/docs/r/storage_mover_agent.html.markdown (hunk 6 above)
Comment (sinbai) [reply website/docs/r/storage_mover_agent.html.markdown]: Fixed.

File: website/docs/r/storage_mover_agent.html.markdown (hunk 7 above)
Comment (sinbai) [reply website/docs/r/storage_mover_agent.html.markdown]: Fixed.

File: website/docs/r/storage_mover_agent.html.markdown (hunk 8 above)
Comment (sinbai) [reply website/docs/r/storage_mover_agent.html.markdown]: Fixed.

File: website/docs/r/storage_mover_agent.html.markdown (hunk 9 above)
Comment (sinbai) [reply website/docs/r/storage_mover_agent.html.markdown]: Maybe the description should be updated as below, refer [here ](https://github.com/Azure/azure-rest-api-specs/blob/5042eea1a5a0471073e37d874687aad62f6cffbb/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2022-11-10/HybridCompute.json#L1293)and [here](https://learn.microsoft.com/en-us/windows-hardware/drivers/bringup/smbios).
image

**Comment by sinbai** [comment]
> hi @sinbai
> 
> Thanks for this PR - I've taken a look through and left some comments inline, if we can fix those up then we should be able to take another look here.
> 
> Thanks!

Hi @tombuildsstuff

Thanks for all your detailed feedback. I have fixed/replied. Could you please take another look?

File: internal/services/storagemover/storage_mover_agent_resource.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+			if model := resp.Model; model != nil {
+				state.ArcVmUuid = model.Properties.ArcVMUuid
+				state.ArcVirtualMachineId = model.Properties.ArcResourceId
+
+				des := ""
+				if model.Properties.Description != nil {
+					des = *model.Properties.Description
+				}
+				state.Description = des
```
Comment (stephybun) [inline internal/services/storagemover/storage_mover_agent_resource.go:201]: We can use the functions from the pointer package to condense this to
```suggestion
				state.Description = pointer.From(model.Properties.Description)
```

File: internal/services/storagemover/storage_mover_agent_resource.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+		},
+
+		"arc_virtual_machine_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: computevalidate.HybridMachineID,
+		},
+
+		"arc_vm_uuid": {
```
Comment (stephybun) [inline internal/services/storagemover/storage_mover_agent_resource.go:57]: For consistency could we spell this out instead of abbreviating it
```suggestion
		"arc_virtual_machine_uuid": {
```

**Review by stephybun — COMMENTED** [review]
Thanks for making those updates @sinbai. I have two suggestions but otherwise this looks like it's good to go!

File: internal/services/storagemover/storage_mover_agent_resource.go (hunk 12 above)
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource.go]: Fixed.

File: internal/services/storagemover/storage_mover_agent_resource.go (hunk 13 above)
Comment (sinbai) [reply internal/services/storagemover/storage_mover_agent_resource.go]: Fixed.

**Comment by sinbai** [comment]
> Thanks for making those updates @sinbai. I have two suggestions and this needs to be rebased but otherwise this looks like it's good to go!

@stephybun Thanks for your feedback. The code has been updated. Could you please take another look?
Test results:
image
